File: app/src/main/java/torille/fi/lurkforreddit/utils/CommentsStreamingParser.java
package torille.fi.lurkforreddit.utils;

import android.support.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import timber.log.Timber;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentResponse;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentChild;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentData;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.MyAdapterFactory;
import torille.fi.lurkforreddit.data.models.jsonResponses.PostDetails;
import torille.fi.lurkforreddit.data.models.jsonResponses.PostResponse;
import torille.fi.lurkforreddit.di.scope.RedditScope;

/**
 * Utility classes to stream parse {@link CommentListing} Json and nested models
 */
@RedditScope
public final class CommentsStreamingParser {

    private static Gson gson;

    @Inject
    public CommentsStreamingParser(@NonNull Gson gson) {
        CommentsStreamingParser.gson = gson;
    }

    /**
     * End point to read more comment api response
     *
     * @param reader needed for parsing
     * @return a list of {@link CommentChild}
     * @throws IOException
     */
    public static List<CommentChild> readMoreComments(JsonReader reader) throws IOException {

        List<CommentChild> comments = new ArrayList<>();
        reader.beginObject();
        while (reader.hasNext()) {
            String name = reader.nextName();
            if (name.equals("json")) {
                comments = readMore(reader, "data");
            } else {
                reader.skipValue();
            }
        }
        reader.endObject();

        return comments;
    }

    /**
     * Reads and parses more comments api response
     *
     * @param reader    used for parsing
     * @param searchFor used to travel the response
     * @return at the end returns a list of  {@link CommentChild}
     * @throws IOException
     */
    private static List<CommentChild> readMore(JsonReader reader, String searchFor) throws IOException {
        List<CommentChild> comments = new ArrayList<>();

        reader.beginObject();
        while (reader.hasNext()) {
            if (reader.nextName().equals(searchFor)) {
                switch (searchFor) {
                    case "json":
                        return readMore(reader, "data");
                    case "data":
                        return readMore(reader, "things");
                    case "things":
                        return readCommentChildren(reader);
                    default:
                        reader.skipValue();
                }
            } else {
                reader.skipValue();
            }

        }
        reader.endObject();
        return comments;
    }

    public List<CommentListing> readCommentListingArray(JsonReader reader) throws IOException {
        List<CommentListing> list = new ArrayList<>();

        reader.beginArray();
        while (reader.hasNext()) {
            list.add(readCommentListing(reader));
        }
        reader.endArray();

        return list;
    }

    private static CommentListing readCommentListing(JsonReader reader) throws IOException {
        String kind = null;
        CommentData commentData = null;

        reader.beginObject();
        while (reader.hasNext()) {
            String name = reader.nextName();
            if (name.equals("kind")) {
                kind = reader.nextString();

            } else if (name.equals("data") && reader.peek() != JsonToken.NULL) {
                commentData = readCommentData(reader);
            } else {
                reader.skipValue();
            }
        }
        reader.endObject();
        return CommentListing.create(kind, commentData);
    }

    private static CommentData readCommentData(JsonReader reader) throws IOException {
        List<CommentChild> commentChildren = new ArrayList<>();
        String after = null;
        String before = null;

        reader.beginObject();
        while (reader.hasNext()) {
            String name = reader.nextName();
            if (name.equals("children") && reader.peek() != JsonToken.NULL) {
                commentChildren = readCommentChildren(reader);
            } else if (name.equals("after") && reader.peek() != JsonToken.NULL) {
                after = reader.nextString();
            } else if (name.equals("before") && reader.peek() != JsonToken.NULL) {
                before = reader.nextString();
            } else {
                reader.skipValue();
            }
        }
        reader.endObject();
        return CommentData.create(commentChildren, after, before);

    }

    private static List<CommentChild> readCommentChildren(JsonReader reader) throws IOException {

        List<CommentChild> commentChildren = new ArrayList<>();
        if (reader.peek() == JsonToken.BEGIN_OBJECT) {
            reader.beginObject();
            while (reader.hasNext()) {
                commentChildren.add(readCommentChild(reader));
            }
            reader.endObject();
        } else if (reader.peek() == JsonToken.BEGIN_ARRAY) {
            reader.beginArray();
            while (reader.hasNext()) {
                commentChildren.add(readCommentChild(reader));
            }
            reader.endArray();
        }
        return commentChildren;

    }

    private static CommentChild readCommentChild(JsonReader reader) throws IOException {
        String kind = null;
        CommentResponse comment = null;
        PostDetails postDetails = null;

        reader.beginObject();
        while (reader.hasNext()) {
            String name = reader.nextName();
            switch (name) {
                case "kind":
                    kind = reader.nextString();
                    break;
                case "data":
                    if (kind != null && kind.equals("t3")) {
                        postDetails = gson.fromJson(reader, PostDetails.class);
                    } else {
                        comment = readComment(reader);
                    }
                    break;
                default:
                    reader.skipValue();
                    break;
            }
        }
        reader.endObject();
        return CommentChild.create(kind, comment, postDetails);

    }

    private static CommentResponse readComment(JsonReader reader) throws IOException {
        String subredditId = null;
        String linkId = null;
        CommentListing replies = null;
        boolean saved = false;
        String id = null;
        int gilded = 0;
        boolean archived = false;
        String author = null;
        String parendId = null;
        int score = 0;
        int controversiality = 0;
        String bodyHtml = null;
        boolean stickied = false;
        String subreddit = null;
        boolean scoreHidden = false;
        String name = null;
        long createdUtc = 0;
        String authorFlairText = null;
        int ups = 0;
        int count = 0;
        List<String> childrenIds = null;

        reader.beginObject();
        while (reader.hasNext()) {
            String jsonName = reader.nextName();
            switch (jsonName) {
                case "subreddit_id":
                    subredditId = reader.nextString();
                    break;
                case "link_id":
                    linkId = reader.nextString();
                    break;
                case "replies":
                    if (reader.peek() == JsonToken.BEGIN_OBJECT) {
                        replies = readCommentListing(reader);
                    } else {
                        reader.skipValue();
                    }
                    break;
                case "saved":
                    saved = reader.nextBoolean();
                    break;
                case "id":
                    id = reader.nextString();
                    break;
                case "gilded":
                    gilded = reader.nextInt();
                    break;
                case "archived":
                    archived = reader.nextBoolean();
                    break;
                case "author":
                    author = reader.nextString();
                    break;
                case "parent_id":
                    parendId = reader.nextString();
                    break;
                case "score":
                    score = reader.nextInt();
                    break;
                case "controversiality":
                    controversiality = reader.nextInt();
                    break;
                case "body_html":
                    bodyHtml = reader.nextString();
                    break;
                case "stickied":
                    stickied = reader.nextBoolean();
                    break;
                case "subreddit":
                    subreddit = reader.nextString();
                    break;
                case "score_hidden":
                    scoreHidden = reader.nextBoolean();
                    break;
                case "name":
                    name = reader.nextString();
                    break;
                case "created_utc":
                    createdUtc = reader.nextLong();
                    break;
                case "author_flair_text":
                    if (reader.peek() != JsonToken.NULL) {
                        authorFlairText = reader.nextString();
                    } else {
                        reader.skipValue();
                    }
                    break;
                case "ups":
                    ups = reader.nextInt();
                    break;
                case "count":
                    count = reader.nextInt();
                    break;
                case "children":
                    if (reader.peek() == JsonToken.BEGIN_ARRAY) {
                        childrenIds = readStringList(reader);
                        break;
                    } else {
                        reader.skipValue();
                        break;
                    }
                default:
                    reader.skipValue();
                    break;
            }
        }
        reader.endObject();
        return CommentResponse.builder()
                .setLinkId(linkId)
                .setReplies(replies)
                .setSaved(saved)
                .setId(id)
                .setGilded(gilded)
                .setArchived(archived)
                .setAuthor(author)
                .setCount(count)
                .setParentId(parendId)
                .setScore(score)
                .setControversiality(controversiality)
                .setBodyHtml(bodyHtml)
                .setStickied(stickied)
                .setSubreddit(subreddit)
                .setScoreHidden(scoreHidden)
                .setName(name)
                .setAuthorFlairText(authorFlairText)
                .setCreatedUtc(createdUtc)
                .setUps(ups)
                .setChildren(childrenIds)
                .build();

    }

    private static List<String> readStringList(JsonReader reader) throws IOException {
        List<String> childrenIds = new ArrayList<>();

        reader.beginArray();
        while (reader.hasNext()) {
            childrenIds.add(reader.nextString());
        }
        reader.endArray();
        return childrenIds;

    }
}

File: app/src/main/java/torille/fi/lurkforreddit/utils/CommentsStreamingParser.kt
package torille.fi.lurkforreddit.utils

import com.google.gson.Gson
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import torille.fi.lurkforreddit.data.models.jsonResponses.*
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Utility classes to stream parse [CommentListing] Json and nested models
 */
@Singleton
class CommentsStreamingParser @Inject
constructor(val gson: Gson) {

    /**
     * End point to read more comment api response

     * @param reader needed for parsing
     * *
     * @return a list of [CommentChild]
     * *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun readMoreComments(reader: JsonReader): List<CommentChild> {
        var comments: List<CommentChild> = emptyList()

        reader.beginObject()
        while (reader.hasNext()) {
            val name = reader.nextName()
            if (name == "json") {
                comments = readMore(reader, "data")
            } else {
                reader.skipValue()
            }
        }
        reader.endObject()

        return comments
    }

    @Throws(IOException::class)
    fun readCommentListingArray(reader: JsonReader): List<CommentListing> {
        var list: List<CommentListing> = emptyList()

        reader.beginArray()
        while (reader.hasNext()) {
            val listing = readCommentListing(reader)
            list += listing
        }
        reader.endArray()

        return list
    }

    /**
     * Reads and parses more comments api response

     * @param reader    used for parsing
     * *
     * @param searchFor used to travel the response
     * *
     * @return at the end returns a list of  [CommentChild]
     * *
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun readMore(reader: JsonReader, searchFor: String): List<CommentChild> {
        val comments: List<CommentChild> = emptyList()

        reader.beginObject()
        while (reader.hasNext()) {
            if (reader.nextName() == searchFor) {
                when (searchFor) {
                    "json" -> return readMore(reader, "data")
                    "data" -> return readMore(reader, "things")
                    "things" -> return readCommentChildren(reader)
                    else -> reader.skipValue()
                }
            } else {
                reader.skipValue()
            }

        }
        reader.endObject()

        return comments
    }

    @Throws(IOException::class)
    private fun readCommentListing(reader: JsonReader): CommentListing {
        var kind = ""
        var commentData: CommentData = CommentData(emptyList(), "", "")

        reader.beginObject()
        while (reader.hasNext()) {
            val name = reader.nextName()
            if (name == "kind") {
                kind = reader.nextString()
            } else if (name == "data" && reader.peek() != JsonToken.NULL) {
                commentData = readCommentData(reader)
            } else {
                reader.skipValue()
            }
        }
        reader.endObject()

        return CommentListing(kind = kind, commentData = commentData)
    }

    @Throws(IOException::class)
    private fun readCommentData(reader: JsonReader): CommentData {
        var commentChildren: List<CommentChild> = emptyList()
        var after = ""
        var before = ""

        reader.beginObject()
        while (reader.hasNext()) {
            val name = reader.nextName()
            if (name == "children" && reader.peek() != JsonToken.NULL) {
                commentChildren += readCommentChildren(reader)
            } else if (name == "after" && reader.peek() != JsonToken.NULL) {
                after = reader.nextString()
            } else if (name == "before" && reader.peek() != JsonToken.NULL) {
                before = reader.nextString()
            } else {
                reader.skipValue()
            }
        }
        reader.endObject()

        return CommentData(commentChildren, after, before)
    }

    @Throws(IOException::class)
    private fun readCommentChildren(reader: JsonReader): List<CommentChild> {
        var commentChildren: List<CommentChild> = emptyList()

        if (reader.peek() == JsonToken.BEGIN_OBJECT) {
            reader.beginObject()
            while (reader.hasNext()) {
                commentChildren += readCommentChild(reader)
            }
            reader.endObject()
        } else if (reader.peek() == JsonToken.BEGIN_ARRAY) {
            reader.beginArray()
            while (reader.hasNext()) {
                commentChildren += readCommentChild(reader)
            }
            reader.endArray()
        }

        return commentChildren
    }

    @Throws(IOException::class)
    private fun readCommentChild(reader: JsonReader): CommentChild {
        var kind = ""
        var postDetails: PostDetails? = null
        var comment = CommentResponse()

        reader.beginObject()
        while (reader.hasNext()) {
            val name = reader.nextName()
            when (name) {
                "kind" -> kind = reader.nextString()
                "data" ->
                    if (kind == "t3") {
                        postDetails = gson.fromJson(reader, PostDetails::class.java)
                    } else {
                        comment = readComment(reader)
                    }
                else -> reader.skipValue()
            }
        }
        reader.endObject()

        return CommentChild(kind, comment, postDetails)
    }

    @Throws(IOException::class)
    private fun readComment(reader: JsonReader): CommentResponse {
        var subredditId = ""
        var linkId = ""
        var replies: CommentListing = CommentListing()
        var saved = false
        var id = ""
        var gilded = 0
        var archived = false
        var author = ""
        var parendId = ""
        var score = 0
        var controversiality = 0
        var bodyHtml = ""
        var stickied = false
        var subreddit = ""
        var scoreHidden = false
        var edited = false
        var name = ""
        var createdUtc: Long = 0
        var authorFlairText = ""
        var ups = 0
        var count = 0
        var childrenIds: List<String> = emptyList()

        reader.beginObject()
        while (reader.hasNext()) {
            val jsonName = reader.nextName()
            when (jsonName) {
                "subreddit_id" -> subredditId = reader.nextString()
                "link_id" -> linkId = reader.nextString()
                "replies" ->
                    if (reader.peek() == JsonToken.BEGIN_OBJECT) {
                        replies = readCommentListing(reader)
                    } else {
                        reader.skipValue()
                    }
                "saved" -> saved = reader.nextBoolean()
                "subId" -> id = reader.nextString()
                "gilded" -> gilded = reader.nextInt()
                "archived" -> archived = reader.nextBoolean()
                "author" -> author = reader.nextString()
                "parent_id" -> parendId = reader.nextString()
                "score" -> score = reader.nextInt()
                "controversiality" -> controversiality = reader.nextInt()
                "body_html" -> bodyHtml = reader.nextString()
                "stickied" -> stickied = reader.nextBoolean()
                "subreddit" -> subreddit = reader.nextString()
                "score_hidden" -> scoreHidden = reader.nextBoolean()
                "edited" ->
                    if (reader.peek() == JsonToken.NUMBER) {
                        edited = true
                        reader.skipValue()
                    } else {
                        edited = reader.nextBoolean()
                    }
                "name" -> name = reader.nextString()
                "created_utc" -> createdUtc = reader.nextLong()
                "author_flair_text" ->
                    if (reader.peek() != JsonToken.NULL) {
                        authorFlairText = reader.nextString()
                    } else {
                        reader.skipValue()
                    }
                "ups" -> ups = reader.nextInt()
                "count" -> count = reader.nextInt()
                "children" -> {
                    if (reader.peek() == JsonToken.BEGIN_ARRAY) {
                        childrenIds += readStringList(reader)
                    } else {
                        reader.skipValue()
                    }
                }
                else -> reader.skipValue()
            }
        }
        reader.endObject()

        return CommentResponse(
            linkId = linkId,
            replies = replies,
            saved = saved,
            id = id,
            gilded = gilded,
            archived = archived,
            author = author,
            count = count,
            parentId = parendId,
            score = score,
            controversiality = controversiality,
            bodyHtml = bodyHtml,
            stickied = stickied,
            subreddit = subreddit,
            scoreHidden = scoreHidden,
            edited = edited,
            name = name,
            authorFlairText = authorFlairText,
            createdUtc = createdUtc,
            ups = ups,
            children = childrenIds,
            subredditId = subredditId
        )
    }

    @Throws(IOException::class)
    private fun readStringList(reader: JsonReader): List<String> {
        var childrenIds: List<String> = emptyList()

        reader.beginArray()
        while (reader.hasNext()) {
            childrenIds += (reader.nextString())
        }
        reader.endArray()

        return childrenIds
    }

}
